File: ts/src/cli/run-cli/emit-pruning-warnings.test.ts
import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";

import type { PipelineRunDetails } from "../../pipeline/runner/pipeline-run-details.js";
import { emitPruningWarnings } from "./emit-pruning-warnings.js";

describe("emitPruningWarnings", () => {
  let writeSpy: ReturnType<typeof vi.spyOn>;
  let written: /* mutable */ string[];

  beforeEach(() => {
    written = [];
    writeSpy = vi
      .spyOn(process.stderr, "write")
      .mockImplementation((chunk: unknown) => {
        written.push(typeof chunk === "string" ? chunk : String(chunk));
        return true;
      });
  });

  afterEach(() => {
    writeSpy.mockRestore();
  });

  test("writes nothing when pruning is null", () => {
    emitPruningWarnings(null);
    expect(writeSpy).not.toHaveBeenCalled();
  });

  test("writes nothing when pruning is empty", () => {
    emitPruningWarnings([]);
    expect(writeSpy).not.toHaveBeenCalled();
  });

  test("writes nothing when every entry has matched > 0", () => {
    const pruning = [
      { query: "render", matched: 3 },
      { query: "init", matched: 1 },
    ] as const satisfies PipelineRunDetails["pruning"];

    emitPruningWarnings(pruning);
    expect(writeSpy).not.toHaveBeenCalled();
  });

  test("writes a warning line for an entry with matched === 0", () => {
    const pruning = [
      { query: "render", matched: 0 },
    ] as const satisfies PipelineRunDetails["pruning"];

    emitPruningWarnings(pruning);
    expect(written).toEqual(["uns: warning: query 'render' matched 0 roots\n"]);
  });

  test("writes one warning line per zero-match entry, skipping the matched ones", () => {
    const pruning = [
      { query: "render", matched: 0 },
      { query: "init", matched: 2 },
      { query: "boot", matched: 0 },
    ] as const satisfies PipelineRunDetails["pruning"];

    emitPruningWarnings(pruning);
    expect(written).toEqual([
      "uns: warning: query 'render' matched 0 roots\n",
      "uns: warning: query 'boot' matched 0 roots\n",
    ]);
  });
});

File: crates/unsnarl/src/cli/run_cli/emit_pruning_warnings_test.rs
use super::*;

fn capture(pruning: Option<&[PrunePerQueryDetail]>) -> String {
    let mut buf = Vec::new();
    emit_pruning_warnings(pruning, &mut buf);
    String::from_utf8(buf).expect("stderr should be valid UTF-8")
}

#[test]
fn writes_nothing_when_pruning_is_none() {
    assert_eq!(capture(None), "");
}

#[test]
fn writes_nothing_when_pruning_is_empty() {
    let empty: Vec<PrunePerQueryDetail> = Vec::new();
    assert_eq!(capture(Some(&empty)), "");
}

#[test]
fn writes_nothing_when_every_entry_has_matched_above_zero() {
    let entries = vec![
        PrunePerQueryDetail {
            query: "render".to_string(),
            matched: 3,
        },
        PrunePerQueryDetail {
            query: "init".to_string(),
            matched: 1,
        },
    ];
    assert_eq!(capture(Some(&entries)), "");
}

#[test]
fn writes_a_warning_line_for_an_entry_with_matched_zero() {
    let entries = vec![PrunePerQueryDetail {
        query: "render".to_string(),
        matched: 0,
    }];
    assert_eq!(
        capture(Some(&entries)),
        "uns: warning: query 'render' matched 0 roots\n"
    );
}

#[test]
fn writes_one_warning_line_per_zero_match_entry_skipping_matched_ones() {
    let entries = vec![
        PrunePerQueryDetail {
            query: "render".to_string(),
            matched: 0,
        },
        PrunePerQueryDetail {
            query: "init".to_string(),
            matched: 2,
        },
        PrunePerQueryDetail {
            query: "boot".to_string(),
            matched: 0,
        },
    ];
    assert_eq!(
        capture(Some(&entries)),
        "uns: warning: query 'render' matched 0 roots\nuns: warning: query 'boot' matched 0 roots\n"
    );
}
